test/unittest/unittest_nntrainer_models.cpp

```diff
@@ -0,0 +1,359 @@
+   *
+   * @param deriv dervatives
+   * @param iteration iteration
+   * @param should_verify should verify the inputs/gradients/outputs
+   * @return nntrainer::sharedConstTensor
+   */
+  nntrainer::sharedConstTensor backward(nntrainer::sharedConstTensor deriv,
+                                        int iteration,
+                                        bool should_verify = true);
+
+  /**
+   * @brief verify weights of the current node
+   *
+   * @param error_msg error_msg to put when fail
+   */
+  void verifyWeight(const std::string &error_msg);
+
+  /**
+   * @brief verify gradient of the current node
+   *
+   * @param error_msg error_msg to put when fail
+   */
+  void verifyGrad(const std::string &error_msg);
+
+  /**
+   * @brief Get the Loss
+   *
+   * @return float loss
+   */
+  float getLoss() { return node->getLoss(); }
+
+  /**
+   * @brief read Node
+   *
+   * @param in input file stream
+   */
+  void read(std::ifstream &in);
+
+private:
+  NodeType node;
+  nntrainer::Tensor expected_input;
+  nntrainer::Tensor expected_output;
+  nntrainer::Tensor expected_dx;
+  std::vector<nntrainer::Weight> expected_weights;
+};
+
+class GraphWatcher {
+public:
+  using WatchedFlatGraph = std::vector<NodeWatcher>;
+  GraphWatcher(const std::string &config);
+
+  void compareFor(const std::string &reference,
+                  const nntrainer::TensorDim &label_shape,
+                  unsigned int iterations);
+
+private:
+  void readIteration(std::ifstream &f);
+
+  nntrainer::NeuralNetwork nn;
+  WatchedFlatGraph nodes;
+  NodeWatcher loss_node;
+  float expected_loss;
+};
+
+void NodeWatcher::read(std::ifstream &in) {
+  expected_input.read(in);
+
+  /// @note below is harrasing the fact the tensor shares same base memory
+  /// it should better be getVariableRef() or somewhat equivalent
+  for (auto &i : expected_weights) {
+    i.getVariable().read(in);
+  }
+
+  for (auto &i : expected_weights) {
+    i.getGradient().read(in);
+  }
+
+  expected_output.read(in);
+  expected_dx.read(in);
+}
+
+void NodeWatcher::verifyWeight(const std::string &error_msg) {
+  for (unsigned int i = 0; i < expected_weights.size(); ++i) {
+    verify(node->weightAt(i).getVariable(), expected_weights[i].getVariable(),
+           error_msg + " " + node->weightAt(i).getName() + "weight");
+  }
+}
+
+void NodeWatcher::verifyGrad(const std::string &error_msg) {
+  for (unsigned int i = 0; i < expected_weights.size(); ++i) {
+    verify(node->weightAt(i).getGradient(), expected_weights[i].getGradient(),
+           error_msg + " " + node->weightAt(i).getName() + "grad");
+  }
+}
+
+nntrainer::sharedConstTensor
+NodeWatcher::forward(nntrainer::sharedConstTensor in, int iteration) {
+  std::stringstream ss;
+  ss << "forward failed at " << node->getName() << " at iteration "
+     << iteration;
+  std::string err_msg = ss.str();
+
+  verify(*in, expected_input, err_msg + " at input ");
+  verifyWeight(err_msg);
```

Comment:
redundant verifyWeight()?

Reply:
This assures that weights are the same as expected before forwarding. (different from after forwarding as it can be modified by forwarding by a wrong implementation)

Reply:
Because we match after updation, this can be deleted I guess. Do you have an opnion about this?

Reply:
Alright I'll update

Reply:
refer to https://github.com/nnstreamer/nntrainer/pull/653/files#r509012565